stoker/stoker.go

```diff
@@ -0,0 +1,145 @@
+// Stoker loads a dag from a remote host into an in-memory blockstore with the
+// intention that it will be processed in a traversal-order compatible way such
+// that data can be efficiently streamed and not fully buffered in memory.
+package stoker
+
+import (
+	"context"
+	"io"
+
+	"github.com/filecoin-saturn/caboose"
+	"github.com/ipld/go-car"
+
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/ipfs/go-cid"
+	ipfsblockstore "github.com/ipfs/go-ipfs-blockstore"
+	"github.com/ipfs/go-libipfs/blocks"
+	golog "github.com/ipfs/go-log/v2"
+)
+
+var goLogger = golog.Logger("caboose/stoker")
+
+type Stoker struct {
+	*caboose.Caboose
+}
+
+func New(c *caboose.Caboose) *Stoker {
+	return &Stoker{c}
+}
+
+const SessionCacheSize = 1024
+
+func (s *Stoker) NewSession(ctx context.Context, path string) ipfsblockstore.Blockstore {
+	cache, err := lru.NewARC(SessionCacheSize)
+	if err != nil {
+		goLogger.Warnw("failed to allocate cache for session", "err", err)
+		return nil
+	}
+
+	ss := stokerSession{
+		c:                      s.Caboose,
+		cache:                  cache,
+		writeAheadSynchronizer: NewWAS(SessionCacheSize*3/4, SessionCacheSize*1/4),
+	}
+	go ss.fillCache(ctx, path)
+
+	return &ss
+}
+
+type stokerSession struct {
+	c     *caboose.Caboose
+	cache *lru.ARCCache
+	*writeAheadSynchronizer
+}
+
+func (ss *stokerSession) fillCache(ctx context.Context, path string) error {
+	err := ss.c.Fetch(ctx, path, func(resource string, reader io.Reader) error {
+		cr, err := car.NewCarReader(reader)
+		if err != nil {
+			return err
+		}
+
+		n := 0
+		for {
+			ss.writeAheadSynchronizer.ReqAdd()
```

Comment:
IIUC, this is to limit reading of incoming blocks from the CAR stream to keep memory usage in check, right ?
Given that the cache size is bounded, I don't think we need this to limit how many blocks we put in the cache.

Reply:
Basically, what would happen if we didn't use a `writeAheadSynchronizer ` here ?